sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala

```diff
@@ -129,20 +138,24 @@ class StreamExecution(
       val optimizerTime = (System.nanoTime() - optimizerStart).toDouble / 1000000
       logDebug(s"Optimized batch in ${optimizerTime}ms")
 
-      // Update the offsets and calculate a new composite offset
-      newOffsets.foreach(currentOffsets.update)
-      val newStreamProgress = logicalPlan.collect {
-        case StreamingRelation(source, _) => currentOffsets.get(source)
+      StreamExecution.this.synchronized {
+        // Update the offsets and calculate a new composite offset
+        newOffsets.foreach(currentOffsets.update)
+        val newStreamProgress = logicalPlan.collect {
+          case StreamingRelation(source, _) => currentOffsets.get(source)
+        }
+        val batchOffset = CompositeOffset(newStreamProgress)
+
+        // Construct the batch and send it to the sink.
+        val nextBatch = new Batch(batchOffset, new DataFrame(sqlContext, newPlan))
+        sink.addBatch(nextBatch)
       }
-      val batchOffset = CompositeOffset(newStreamProgress)
 
-      // Construct the batch and send it to the sink.
-      val nextBatch = new Batch(batchOffset, new DataFrame(sqlContext, newPlan))
-      sink.addBatch(nextBatch)
+      awaitBatchLock.synchronized {
```

Comment:
note regarding offline discussion: confusing to have volatile batchRun in synchronized, while other accesses of batchRun is not in synchronized.